sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py

```diff
@@ -107,17 +192,124 @@ def __init__(
+        try:
+            self._buffered_producer_dispatcher.enqueue_events(events, **kwargs)
+        except AttributeError:
+            self._get_partitions()
+            self._get_max_message_size()
+            self._buffered_producer_dispatcher = BufferedProducerDispatcher(
+                self._partition_ids,
+                self._on_success,
+                self._on_error,
+                self._create_producer,
+                self.eventhub_name,
+                self._max_message_size_on_link,
+                max_wait_time=self._max_wait_time,
+                max_buffer_length=self._max_buffer_length,
+                executor=self._executor,
+                amqp_transport=self._amqp_transport
+            )
+            self._buffered_producer_dispatcher.enqueue_events(events, **kwargs)
+
+    def _batch_preparer(self, event_data_batch, **kwargs):
+        partition_id = kwargs.pop("partition_id", None)
+        partition_key = kwargs.pop("partition_key", None)
+
+        if isinstance(event_data_batch, EventDataBatch):
+            if partition_id or partition_key:
+                raise TypeError(
+                    "partition_id and partition_key should be None when sending an EventDataBatch "
+                    "because type EventDataBatch itself may have partition_id or partition_key"
+                )
+            to_send_batch = event_data_batch
+        else:
+            to_send_batch = self.create_batch(
+                partition_id=partition_id, partition_key=partition_key
+            )
+            to_send_batch._load_events( # pylint:disable=protected-access
+                event_data_batch
+            )
+
+        return (
+            to_send_batch,
+            to_send_batch._partition_id,    # pylint:disable=protected-access
+            partition_key,
+        )
+
+    def _buffered_send_batch(self, event_data_batch, **kwargs):
+        batch, pid, pkey = self._batch_preparer(event_data_batch, **kwargs)
+
+        if len(batch) == 0:
+            return
+
+        timeout = kwargs.get("timeout")
+        timeout_time = time.time() + timeout if timeout else None
+
+        self._buffered_send(
+            event_data_batch,
+            partition_id=pid,
+            partition_key=pkey,
+            timeout_time=timeout_time,
+        )
+
+    def _buffered_send_event(self, event, **kwargs):
+        partition_key = kwargs.get("partition_key")
+        set_event_partition_key(event, partition_key, self._amqp_transport)
+        timeout = kwargs.get("timeout")
+        timeout_time = time.time() + timeout if timeout else None
+        self._buffered_send(
+            event,
+            partition_id=kwargs.get("partition_id"),
+            partition_key=partition_key,
+            timeout_time=timeout_time,
+        )
+
     def _get_partitions(self):
         # type: () -> None
         if not self._partition_ids:
-            _LOGGER.debug("Populating partition IDs so producers can be started.")
```

Comment:
I decided to move them from info level to debug, so that if the issue arises in the future where we want to see which path send ended up taking, it would be helpful. That line would be handy in situations where the producer was open, then closed and then reopened and something went wonky after wards.

My choice is to keep them there :)

Reply:
leaving in.